core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -115,11 +123,29 @@ public void clusterChanged(ClusterChangedEvent event) {
             return;
         }
 
-        for (IndexRoutingTable indexRoutingTable : event.state().routingTable()) {
+        RoutingTable routingTable = event.state().routingTable();
+
+        // remove entries from cache that don't exist in the routing table anymore
+        for (Iterator<ShardId> it = deletionSkipCache.iterator(); it.hasNext(); ) {
+            ShardId shardId = it.next();
+            if (routingTable.hasIndex(shardId.getIndex()) == false) {
+                it.remove();
+            }
+        }
+        // remove entries from cache which are allocated to this node
+        final String localNodeId = event.state().nodes().getLocalNodeId();
+        RoutingNode localRoutingNode = event.state().getRoutingNodes().node(localNodeId);
+        if (localRoutingNode != null) {
```

Comment:
I'm wondering if we should make IndicesStore be called from IndicesClusterStateService rather than making it a generic listener? As follow up of course.